Architecture/redesign

MatHazak opened the pull request:
I tried to redesign the architecture by repackaging, moving some components and implementing some new components.

oryanm commented:
This looks like a big change. 
Please list out what this redesign consists of, and the reasoning behind every change.

MatHazak commented:
The main goal was to separate data-related operations from the UI layer. I implemented MVVM architecture using ViewModels. Moving the response handling (including JSON parsing) into the TrelloApi class and making requests through the repository, where we will introduce local database in the future and making the data supply consistence and clean.

I just added a description for each commit. Feel free to ask if anything is missing.

MatHazak commented:
@oryanm 
Would you please make a response? I'm waiting on this.
As time goes by, I have less and less time to spend on this project. I don't want to rush and do things without enough attention, but I think it's taking more time than it needs.

oryanm commented on app/src/main/kotlin/com/github/oryanmat/trellowidget/util/network/VolleyUtil.kt:
```diff
@@ -0,0 +1,58 @@
+import com.github.oryanmat.trellowidget.util.Json
+import java.lang.reflect.Type
+
+class VolleyUtil(appContext: Context) {
```
I don't think this class makes sense. Unless we're planning on adding more APIs (I don't think so?) these are just private methods in `TrelloApi`

MatHazak replied:
It's more of a personal preference, imo. I think this separation helps keep the code organized and maintainable. But if we prefer `TrelloApi` compact, I'm fine with merging this class.

oryanm commented on app/src/main/kotlin/com/github/oryanmat/trellowidget/util/network/DataStatus.kt:
```diff
@@ -0,0 +1,13 @@
+package com.github.oryanmat.trellowidget.util.network
+
+data class DataStatus<out T>(val status: Status, val data: T? = null, val msg: String? = null) {
```
This name is meh. This is more of a `ResponseEntity` or just `Response`. 

Also consider making it a sealed interface with 2 implementations. That would make expressing the difference between success and failure cleaner.

MatHazak replied:
> This name is meh. This is more of a `ResponseEntity` or just `Response`.

I would suggest `ApiResponse`.

> Also consider making it a sealed interface with 2 implementations.

Good point. I will do the implementation soon.

oryanm reviewed (commented):
So my understanding this is just a refactor that doesn't change functionality, yes?
This looks fine to me overall

MatHazak commented on app/src/main/kotlin/com/github/oryanmat/trellowidget/data/remote/ApiResponse.kt:
```diff
@@ -0,0 +1,9 @@
+package com.github.oryanmat.trellowidget.data.remote
+
+
+sealed class ApiResponse<out T> {
+
+    data class Success<out T>(val data: T) : ApiResponse<T>()
+
+    data class Error<out T>(val error: String) : ApiResponse<T>()
```
Because of the type erasure in Java, we can't conveniently use a sealed interface with generic functions. Instead, I used a generic sealed class for this.

oryanm replied:
The limitation regarding generics is true for an abstract class the same way it is true for an interface. 
So why not use an interface?

MatHazak commented:
> So my understanding this is just a refactor that doesn't change functionality, yes?

From a user perspective, yes, but there are actually new components with new functionality.

MatHazak commented:
It feels like we're stuck here.
I'm going to merge it into my fork and move on. I will work on #7.

oryanm reviewed (approved):
This looks good. We can merge it.
